<a href="https://colab.research.google.com/github/setyo-dwi-pratama/FGA-DTS-KOMINFO/blob/main/pertemuan13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Persiapan Dataset**

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

In [3]:
df = pd.read_csv(url, header=None)
df.head()

0    1            2    3    4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450

[5 rows x 26 columns]

In [4]:
headers = ['symboling','normalized-losses','make','fuel-type','aspiration','num-of-doors','body-style',
           'drive-wheels','engine-location','wheel-base','length','width','height','curb-weight','engine-type',
           'num-of-cylinders','engine-size','fuel-system','bore','stroke','compression-ratio','horspower','peak-rpm',
           'city-mpg','highway-mpg','price']

In [5]:
df.columns=headers
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

In [7]:
# Menggantikan nilai pada dataset dari '?' dengan NaN, sehingga terbaca sebagai missing value
df.replace("?", np.nan, inplace=True)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [8]:
# Melihat missing value berdasarkan True dan False
missing_data = df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  ...  city-mpg  highway-mpg  price
0      False               True  False  ...     False        False  False
1      False               True  False  ...     False        False  False
2      False               True  False  ...     False        False  False
3      False              False  False  ...     False        False  False
4      False              False  False  ...     False        False  False

[5 rows x 26 columns]

In [9]:
# Melihat kolom-kolom yang terdapat missing value
for column in missing_data.columns.values.tolist():
  print("Nama Kolom", column)
  print(missing_data[column].value_counts())
  print("")

Nama Kolom symboling
False    205
Name: symboling, dtype: int64

Nama Kolom normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

Nama Kolom make
False    205
Name: make, dtype: int64

Nama Kolom fuel-type
False    205
Name: fuel-type, dtype: int64

Nama Kolom aspiration
False    205
Name: aspiration, dtype: int64

Nama Kolom num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

Nama Kolom body-style
False    205
Name: body-style, dtype: int64

Nama Kolom drive-wheels
False    205
Name: drive-wheels, dtype: int64

Nama Kolom engine-location
False    205
Name: engine-location, dtype: int64

Nama Kolom wheel-base
False    205
Name: wheel-base, dtype: int64

Nama Kolom length
False    205
Name: length, dtype: int64

Nama Kolom width
False    205
Name: width, dtype: int64

Nama Kolom height
False    205
Name: height, dtype: int64

Nama Kolom curb-weight
False    205
Name: curb-weight, dtype: int64

Nama Kolom engine-type
False    205
Name: en

Sehingga dari 205 baris terdapat 7 kolom yang memiliki missing value, diantaranya:
1. "normalized-losses": 41 missing data
2. "num-of-doors": 2 missing data
3. "bore": 4 missing data
4. "stroke" : 4 missing data
5. "horsepower": 2 missing data
6. "peak-rpm": 2 missing data
7. "price": 4 missing data

# **Cleaning Dataset**

## **Kolom normalized-losses** 
Mengganti dengan nilai rata-rata

In [30]:
#Konversi type data float
df['normalized-losses'] = df['normalized-losses'].astype(float)

In [31]:
rata = df['normalized-losses'].mean()

In [32]:
rata

122.0

In [33]:
df['normalized-losses'] = df['normalized-losses'].replace(np.nan, rata)

In [34]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom num-of-doors**

In [35]:
df['num-of-doors'].isnull().sum()

0

In [36]:
# Menggunakan fungsi idxmax() untuk kita melihat nilai yang umum atau sering muncul sebagai dasar pengganti pada missing value pada kolom num-of-doors
df['num-of-doors'].value_counts().idxmax()

'four'

In [37]:
df['num-of-doors'].replace(np.nan, "four", inplace=True)

In [38]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom bore**

In [39]:
df['bore'].isnull().sum()

4

In [40]:
#Konversi type data float
df['bore'] = df['bore'].astype(float)

In [41]:
rata1 = df['bore'].mean()
rata1

3.3297512437810957

In [42]:
df['bore'] = df['bore'].replace(np.nan, rata1)

In [43]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom stroke**

In [45]:
df['stroke'].isnull().sum()

4

In [46]:
#Konversi type data float
df['stroke'] = df['stroke'].astype(float)

In [47]:
rata2 = df['stroke'].mean()
rata2

3.2554228855721337

In [48]:
df['stroke'] = df['stroke'].replace(np.nan, rata2)

In [49]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom horspower**

In [51]:
df['horspower'].isnull().sum()

2

In [66]:
#Konversi type data float
df['horspower'] = df['horspower'].astype(float)

In [67]:
rata3 = df['horspower'].mean()
rata3

104.25615763546799

In [69]:
rata33 = int(rata3)
rata33

104

In [70]:
df['horspower'] = df['horspower'].replace(np.nan, rata33)

In [73]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom peak-rpm**

In [74]:
df['peak-rpm'].isnull().sum()

2

In [75]:
#Konversi type data float
df['peak-rpm'] = df['peak-rpm'].astype(float)

In [77]:
rata4 = df['peak-rpm'].mean()
rata4

5125.369458128079

In [78]:
df['peak-rpm'] = df['peak-rpm'].replace(np.nan, rata4)

In [81]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg  price
0            3              122.0  alfa-romero  ...       21          27  13495
1            3              122.0  alfa-romero  ...       21          27  16500
2            1              122.0  alfa-romero  ...       19          26  16500
3            2              164.0         audi  ...       24          30  13950
4            2              164.0         audi  ...       18          22  17450
..         ...                ...          ...  ...      ...         ...    ...
200         -1               95.0        volvo  ...       23          28  16845
201         -1               95.0        volvo  ...       19          25  19045
202         -1               95.0        volvo  ...       18          23  21485
203         -1               95.0        volvo  ...       26          27  22470
204         -1               95.0        volvo  ...       19          25  22625

[205 rows x 26 columns]

## **Kolom price**

In [90]:
df['price'].isnull().sum()

4

In [92]:
#Karena pada kasus ini kolom price merupakan kolom target, maka dari itu pada kolom price yang tidak mempunyai nilai kita akan hapus
df = df.dropna()

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    float64
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

In [95]:
df

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0            3              122.0  alfa-romero  ...       21          27  13495.0
1            3              122.0  alfa-romero  ...       21          27  16500.0
2            1              122.0  alfa-romero  ...       19          26  16500.0
3            2              164.0         audi  ...       24          30  13950.0
4            2              164.0         audi  ...       18          22  17450.0
..         ...                ...          ...  ...      ...         ...      ...
200         -1               95.0        volvo  ...       23          28  16845.0
201         -1               95.0        volvo  ...       19          25  19045.0
202         -1               95.0        volvo  ...       18          23  21485.0
203         -1               95.0        volvo  ...       26          27  22470.0
204         -1               95.0        volvo  ...       19          25  22625.0

[201 rows x 26 columns]

## kita telah membersihkan dataset tanpa missing value dan semua data dalam type data yang sesuai.

# **Standarisasi Data**